pkg/kallsyms/kallsyms.go

```
@@ -127,9 +131,97 @@ func (k *KAllSyms) LookupByInstructionPointer(ip uint64) string {
+	return err == nil
+}
+
+// KernelSymbolAddress looks up a symbol from /proc/kallsyms.
+//
+// symbol: the symbol to lookup
+//
+// Returns the address of the symbol and its module if any
+func KernelSymbolAddress(symbol string) (uint64, string, error) {
+	file, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return 0, "", err
+	}
+	defer file.Close()
+
+	return kernelSymbolAddressFromReader(file, symbol)
+}
+
+// kernelSymbolAddressFromReader
+func kernelSymbolAddressFromReader(r io.Reader, symbol string) (uint64, string, error) {
+	var (
+		count int
+		addr  uint64
+		typ   rune
+		mod   string
+		name  string
+	)
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		// Lines looks like:
+		// ffffffff906e97e0 T security_bprm_check
+		// ffffffffc1cb5010 T netem_module_init    [sch_netem]
+
+		line := scanner.Bytes()
+		typeIdx := bytes.IndexByte(line, byte(' '))
+		if typeIdx == -1 || typeIdx+1 >= len(line) {
+			return 0, "", fmt.Errorf("parsing line %q: no type", line)
+		}
+		nameIdx := bytes.IndexByte(line[typeIdx+1:], byte(' '))
+		if nameIdx == -1 || nameIdx+1+nameIdx+1 >= len(line) {
+			return 0, "", fmt.Errorf("parsing line %q: no symbol", line)
+		}
+		modIdx := bytes.IndexByte(line[typeIdx+1+nameIdx+1:], byte(' '))
+		var nameBytes []byte
+		if modIdx == -1 {
+			// no module
+			nameBytes = line[typeIdx+1+nameIdx+1:]
+			if !bytes.Equal([]byte(symbol), nameBytes) {
+				continue
+			}
+		} else {
+			// module found
+			nameBytes = line[typeIdx+1+nameIdx+1 : typeIdx+1+nameIdx+1+modIdx]
+			if !bytes.Equal([]byte(symbol), nameBytes) {
+				continue
+			}
+		}
+
+		text := scanner.Text()
+		_, err := fmt.Sscanf(text, "%x %c %s\t%s", &addr, &typ, &name, &mod)
+		if err == nil {
+			// a module was found
+			mod = strings.Trim(mod, "[]")
+		} else {
+			// no module was found
+			_, err = fmt.Sscanf(text, "%x %c %s", &addr, &typ, &name)
+		}
+		if err != nil {
+			return 0, "", fmt.Errorf("parsing line %q: %w", text, err)
+		}
+		count++
```

I don't know if this is a worth optimization, but what about testing if count > 1 here to avoid doing uneeded work?

---

It actually saves a lot of time (according to my tests on https://github.com/inspektor-gadget/inspektor-gadget/pull/3871#discussion_r1913325535), specially for symbols that are not in a module because they are located at the beginning of the file `/proc/kallsyms`.

However, cilium/ebpf chose to return an error on duplicate symbols to avoid attaching to the wrong symbol:
https://github.com/cilium/ebpf/blob/9f201154004369b0bc1b99a06ee578aa46c6aab4/internal/kallsyms/kallsyms.go#L121-L123

We are in a different context here since we don't use our kallsyms package to attach to a kernel symbol. For the current usage of `kallsyms.SymbolExists`, it would be a good optimization. But since I am planning to offer the feature to WASM modules where I don't know in advance if it is correct to silently accept ambiguous symbols, I thought it is better to be careful. We can revisit if we need further optimizations.

---

It makes sense! Updated.